### PR TITLE
[PROTOTYPE] Moving to svelte store for explorer

### PR DESCRIPTION
**Pull request by AdityaHegde**

### CODE REVIEW

File: src/lib/application-state-stores/explorer-stores.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
+import { writable } from "svelte/store";
+import type { MetricsExplorerEntity } from "$lib/redux-store/explore/explore-slice";
+import type { MetricViewMetaResponse } from "$common/rill-developer-service/MetricViewActions";
+import type { TimeSeriesTimeRange } from "$common/database-service/DatabaseTimeSeriesActions";
+
+export interface MetricsExplorerStoreType {
+  entities: Record<string, MetricsExplorerEntity>;
+}
+export const MetricsExplorerStore = writable({
```
Comment (hamilton) [inline src/lib/application-state-stores/explorer-stores.ts:9]: I would try to break this out a bit. We want to remove the possibility that the consumer tries to set this store, which they currently can. How I tend to do this:

```
const { set, update, subscribe } = writable(...);
/** code logic etc. etc. *
export const MetricsExplorerStore = { subscribe };
```

All a svelte store needs to be coded as such is a `subscribe` method.

Comment (AdityaHegde) [reply src/lib/application-state-stores/explorer-stores.ts:9]: Ya this feels better. Lemme update the PR

File: src/lib/application-state-stores/explorer-stores.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+      leaderboardMeasureId: meta.measures[0]?.id,
+      filters: {
+        include: [],
+        exclude: [],
+      },
+    })
+  );
+};
+
+export const setMetricsExplorerLeaderboardMeasureId = (
```
Comment (ericpgreen2) [inline src/lib/application-state-stores/explorer-stores.ts:58]: Yeah, like Hamilton mentioned, I think a [custom store](https://svelte.dev/tutorial/custom-stores) with methods to set attributes would be better, e.g. `metricsExplorer.setLeaderboardMeasureId()`

File: src/lib/application-state-stores/explorer-stores.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        state.entities[id] = absenceCallback();
+      }
+      return state;
+    }
+    callback(state.entities[id]);
+    return state;
+  });
+};
+
+export const syncMetricsExplorer = (
```
Comment (ericpgreen2) [inline src/lib/application-state-stores/explorer-stores.ts:30]: This `sync` code is temporary, right? We'll remove it later?

File: src/lib/application-state-stores/explorer-stores.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
+import { writable } from "svelte/store";
+import type { MetricsExplorerEntity } from "$lib/redux-store/explore/explore-slice";
```
Comment (ericpgreen2) [inline src/lib/application-state-stores/explorer-stores.ts:2]: How about we move this type into this file now?

File: src/routes/_surfaces/workspace/explore/time-series-charts/MetricsTimeSeriesCharts.svelte [hunk 5]
```diff
@@ -80,12 +79,12 @@
   let timeSeriesQueryKey = getMetricViewTimeSeriesQueryKey(metricsDefId);
   let timeSeriesQueryFn = () =>
     getMetricViewTimeSeries(metricsDefId, {
-      measures: $metricsExplorer.measureIds,
-      filter: $metricsExplorer.filters,
+      measures: $queryResult.data.measures.map((measure) => measure.id),
```
Comment (ericpgreen2) [inline src/routes/_surfaces/workspace/explore/time-series-charts/MetricsTimeSeriesCharts.svelte:82]: Shouldn't these measures come from the client-side store (`metricsExplorer.selectedMeasures`) not from the server?

File: src/routes/_surfaces/workspace/explore/ExploreHeader.svelte [hunk 6]
```diff
@@ -4,29 +4,45 @@
...
+    getMetricViewMetadata(metricsDefId)
+  );
+  $: {
+    queryKey = getMetricViewMetaQueryKey(metricsDefId);
+    queryResult.setOptions(queryKey, () => getMetricViewMetadata(metricsDefId));
+  }
+  $: syncMetricsExplorer(metricsDefId, $queryResult.data);
+  $: console.log($queryResult?.data);
+
+  $: anythingSelected = isAnythingSelected(metricsExplorer?.filters);
```
Comment (ericpgreen2) [inline src/routes/_surfaces/workspace/explore/ExploreHeader.svelte:42]: More explicit naming could be `$: filtered = isFiltered(...)`

File: src/lib/application-state-stores/explorer-stores.ts [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+import type { MetricViewMetaResponse } from "$common/rill-developer-service/MetricViewActions";
+import type { TimeSeriesTimeRange } from "$common/database-service/DatabaseTimeSeriesActions";
+
+export interface MetricsExplorerStoreType {
+  entities: Record<string, MetricsExplorerEntity>;
```
Comment (ericpgreen2) [inline src/lib/application-state-stores/explorer-stores.ts:7]: Elsewhere in the application we use Contexts to pass entityID-specific stores around, e.g.: https://github.com/rilldata/rill-developer/blob/a2a8e052ac05210da26ef6856b13512f1b2c611e/src/routes/__layout.svelte#L36

Do you like the dictionary-based approach (`Record<string, MetricssExplorerEntity>`) more than the context approach? I googled around a bit for best practices, but couldn't find a consensus. Happy to go with this approach if you prefer it – I just want to be conscious about deviating from our other Context-based methodology.

File: src/routes/_surfaces/workspace/explore/time-series-charts/MetricsTimeSeriesCharts.svelte (hunk 5 above)
Comment (AdityaHegde) [reply src/routes/_surfaces/workspace/explore/time-series-charts/MetricsTimeSeriesCharts.svelte:82]: The meta returns valid measures. Client side can have invalid ones as well. Just wanted to move to a pattern here that always calls the backend.

File: src/lib/application-state-stores/explorer-stores.ts (hunk 4 above)
Comment (AdityaHegde) [reply src/lib/application-state-stores/explorer-stores.ts:2]: Lets remove the redux stuff and do this in there. Didnt touch it in this PR intentionally.

File: src/lib/application-state-stores/explorer-stores.ts (hunk 3 above)
Comment (AdityaHegde) [reply src/lib/application-state-stores/explorer-stores.ts]: We will need to make sure we remove measures and dimensions that is in the store. The sync will take care of it.

File: src/lib/application-state-stores/explorer-stores.ts (hunk 1 above)
Comment (AdityaHegde) [reply src/lib/application-state-stores/explorer-stores.ts]: Done

File: src/lib/application-state-stores/explorer-stores.ts (hunk 7 above)
Comment (AdityaHegde) [reply src/lib/application-state-stores/explorer-stores.ts:7]: Hmmm not sure what the context gives us over just import/export of the store. I feel this is simpler.

File: src/routes/_surfaces/workspace/explore/time-series-charts/MetricsTimeSeriesCharts.svelte (hunk 5 above)
Comment (AdityaHegde) [reply src/routes/_surfaces/workspace/explore/time-series-charts/MetricsTimeSeriesCharts.svelte:82]: Actually you are right.. We need to filter measures from server based on selectedMeasures.

Comment (AdityaHegde) [reply src/routes/_surfaces/workspace/explore/time-series-charts/MetricsTimeSeriesCharts.svelte]: Done

File: src/lib/application-state-stores/explorer-stores.ts [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+      }
+      return state;
+    }
+    callback(state.entities[id]);
+    return state;
+  });
+};
+
+const Methods = {
+  sync(id: string, meta: MetricViewMetaResponse) {
```
Comment (ericpgreen2) [inline src/lib/application-state-stores/explorer-stores.ts:31]: It looks like this `sync` logic has to do with selecting a `leaderboardMeasureId`. Shouldn't this logic go in the `LeaderboardMeasureSelector.svelte` component when that component fetches `/meta`, and we can remove this `sync` function?

If not, is there a better name than `sync`? I'm still a bit confused why it exists.

Comment (ericpgreen2) [reply src/lib/application-state-stores/explorer-stores.ts:31]: If it has something to do with measures getting updated, then, when that happens, the `svelte-query` approach would be to invalidate the `/meta` query and refetch

File: src/lib/application-state-stores/explorer-stores.ts [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+import type { TimeSeriesTimeRange } from "$common/database-service/DatabaseTimeSeriesActions";
+
+export interface MetricsExplorerStoreType {
+  entities: Record<string, MetricsExplorerEntity>;
+}
+const { update, subscribe } = writable({
+  entities: {},
+} as MetricsExplorerStoreType);
+
+const UpdateMetricsExplorer = (
```
Comment (ericpgreen2) [inline src/lib/application-state-stores/explorer-stores.ts:13]: Nit: what's with the capitalization for `UpdateMetricsExplorer`, `Methods`, and `MetricsExplorerStore`? Is that a convention for stores? We don't do it elsewhere AFAIK

File: src/lib/application-state-stores/explorer-stores.ts (hunk 8 above)
Comment (AdityaHegde) [reply src/lib/application-state-stores/explorer-stores.ts:31]: Hmmm we can revisit this in a later PR i feel. This is a quick and easy way to get the measures working.

**Pull request merged**